hibernate-core/src/main/java/org/hibernate/loader/entity/CacheEntityLoaderHelper.java

```diff
@@ -62,6 +62,46 @@
 	private CacheEntityLoaderHelper() {
 	}
 
+	@Incubating
+	public PersistenceContextEntry loadFromSessionCache(
+			EntityKey keyToLoad,
+			LoadEventListener.LoadType options,
+			LockOptions lockOptions,
+			SharedSessionContractImplementor session) {
+		final Object old = session.getEntityUsingInterceptor( keyToLoad );
+
+		if ( old != null ) {
+			// this object was already loaded
+			EntityEntry oldEntry = session.getPersistenceContext().getEntry( old );
+			if ( options.isCheckDeleted() ) {
+				Status status = oldEntry.getStatus();
+				if ( status == Status.DELETED || status == Status.GONE ) {
+					LoadingLogger.LOGGER.debug(
+							"Load request found matching entity in context, but it is scheduled for removal; returning null" );
+					return new PersistenceContextEntry( old, EntityStatus.REMOVED_ENTITY_MARKER );
+				}
+			}
+			if ( options.isAllowNulls() ) {
+				final EntityPersister persister = session.getFactory()
+						.getRuntimeMetamodels()
+						.getMappingMetamodel()
+						.getEntityDescriptor( keyToLoad.getEntityName() );
+				if ( ! persister.isInstance( old ) ) {
+					LOG.debugf(
+							"Load request found matching entity in context, but the matched entity was of an inconsistent return type.  " +
+									"Setting status as `%s`",
+							EntityStatus.INCONSISTENT_RTN_CLASS_MARKER
```

Comment:
We might well add log level checking here for the log message involves string interpolation, so we are gonna pay the cost regardless of whether debug level is enabled or not.